aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -308,5 +310,18 @@ Tensor& logical_xor_out(Tensor& result, const Tensor& self, Scalar other) { retu
 Tensor logical_xor(const Tensor& self, Scalar other) { return comparison_op(self, other, logical_xor_stub); }
 Tensor& logical_xor_(Tensor& self, Scalar other) { return comparison_op_(self, other, logical_xor_stub); }
 
+Tensor& logical_and_out(Tensor& result, const Tensor& self, const Tensor& other) { return comparison_op_out(result, self, other, logical_and_stub); }
```

Comment:
Hi @xuhdev I think I found the root cause why these `logical_*` ops are not working for XLA. 

The problem is in the implementation of `comparison_op` that they should be implemented similar to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UnaryOps.cpp#L52. The current implementation directly call into TensorIterator through stubs which didn't give a chance for device dispatch to get in.
(XLA will be called through logical_xor_out :)

Reply:
Thanks. I think you actually referred to calling `at::` instead of `native::`? I've pushed a change for testing (eventually we should move the change out to a new PR). Interestingly, the comparison operators (ge, le, ...) are not skipped in the XLA CI.

Reply:
@xuhdev `at::comparison_op` won't work because you only have it in `at::native` namespace. But if you pass in the function `out_impl` instead of stub it should work. 
`template <typename OutImpl>
static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {`

Reply:
Yea in `ge`s case it's different sorry it's indeed confusing. 
So `ge` is a non-composable op that XLA is called at `ge` level. But in `logical_*` XLA is expected to be called at `logical_*_out` level. We can change that but need to be a bit careful. 
I was going to write a wiki to explain this better for our OSS contributors, but a quick reference: https://gist.github.com/ailzhang/905aa41a5be8c5315ba0e3d8f0d0e326 contains all currently non-composable ops in Pytorch codebase. which means XLA will be called at these ops.

Reply:
Thanks, this info is indeed helpful. Is there going to be a plan to make all ops either composable or non-composable? I'm just trying to understand what the ultimate design is :)

Reply:
No we'll keep both I think since for non-composable ops autograd is handling them to save a lot of manual work, and that's indeed helpful. 
But we'll definitely look into ways to make this easier to tell for OSS contributors. Another easy way to judge is to look at `build/aten/src/ATen/Declarations.yaml` and check where `abstract=true` for the functions. If you have ideas what could help you understand, we're happy to hear your thoughts as well!